content/src/storage/FileSystemContentStorage.ts

```diff
@@ -83,6 +84,14 @@ export class FileSystemContentStorage implements ContentStorage {
   }
 
   private getFilePath(id: string): string {
-    return path.join(this.root, id)
+    // We are sharding the files using the first 4 digits of its sha1 hash, because it makes troubles
+    // for the file system to handle millions of files in the same directory.
+    // This way, asuming that sha1 hash distribution is ~uniform we are reducing by 16^4 the max amount of files in a directory.
+    const directoryPath = path.join(this.root, createHash('sha1').update(id).digest('hex').substring(0, 4))
+    if (!fs.existsSync(directoryPath)) {
+      fs.mkdirSync(directoryPath, { recursive: true })
+    }
+    const filePath = path.join(directoryPath, id)
+    return filePath
```

Comment:
How does compressed files work?

```
hash(file+extension)[0..3] + / + file
```
or
```
hash(file)[0..3] + / + file + extension
```

Reply:
Option b: `hash(file)[0..3] + / + file + extension`

Reply:
The compression function just takes the file path of the file and save the compressed file in the same filepath + '.gzip'.
So if the `storeStreamAndCompress` method is called, it:
- Stores the uncompressed file in `this.getFilePath(id)`
- It tries to compress the file, if it success, then it places the compressed file in `this.getFilePath(id) + .gzip` (so it is in the same directory as the uncompressed file`
- Then deletes the uncompressed file in `this.getFilePath(id)`

Reply:
Then I don't think this may work with compressed files. Please consider adding integration tests for it:

1. Write file to storage
2. Retrieve file must have encoding null
3. Compress stored file
4. 1) Retrieve file must have encoding gzip 2) Original file (uncompressed) must not exist anymore

Reply:
I ment integration because it does not mock anything. Simply using the `FileSystemContentStorage` and real files.

.storeStreamAndCompress can be used safely for that test, I think.

If the unit test already uses the real FS without failing then we are covered